storage-dynamodb/src/main/java/zipkin2/storage/dynamodb/DynamoDBSpanConsumer.java

```diff
@@ -0,0 +1,332 @@
+  }
+
+  @Override public Call<Void> accept(List<Span> list) {
+    List<WriteRequest> spanWriteRequests = new ArrayList<>(createWriteSpans(list));
+
+    while (!spanWriteRequests.isEmpty()) {
+      int maxIndex = Math.min(spanWriteRequests.size(), 25);
+      BatchWriteItemRequest request = new BatchWriteItemRequest();
+      request.addRequestItemsEntry(spansTableName, spanWriteRequests.subList(0, maxIndex));
+
+      dynamoDB.batchWriteItem(request);
+      spanWriteRequests.subList(0, maxIndex).clear();
+    }
+
+    List<UpdateItemRequest> names = createUpdateNames(list);
+    for (UpdateItemRequest request : names) {
+      dynamoDB.updateItem(request.withTableName(serviceSpanNamesTableName));
+    }
+
+    List<UpdateItemRequest> autocompleteTags = createAutocompleteTags(list);
+    for (UpdateItemRequest request : autocompleteTags) {
+      dynamoDB.updateItem(request.withTableName(autocompleteTagsTableName));
+    }
+
+    return Call.create(null);
+  }
+
+  private List<WriteRequest> createWriteSpans(List<Span> spans) {
+    List<WriteRequest> result = new ArrayList<>(spans.size());
+    for (Span span : spans) {
+      PutRequest spanPut = new PutRequest();
+      spanPut.addItemEntry(TRACE_ID, new AttributeValue(span.traceId()));
+      spanPut.addItemEntry(TRACE_ID_64,
+          new AttributeValue(span.traceId().substring(Math.max(0, span.traceId().length() - 16))));
+
+      spanPut.addItemEntry(TIMESTAMP_SPAN_ID,
+          new AttributeValue().withN(
+              DynamoDBStorage.timestampId(span.timestampAsLong() / 1000, span.id())));
+
+      spanPut.addItemEntry(SPAN_ID, new AttributeValue().withS(span.id()));
+
+      if (span.name() != null && !span.name().isEmpty()) {
+        spanPut.addItemEntry(SPAN_NAME, new AttributeValue().withS(span.name()));
+
+        if (span.localServiceName() != null && !span.localServiceName().isEmpty()) {
+          spanPut.addItemEntry(LOCAL_SERVICE_NAME_SPAN_NAME,
+              new AttributeValue().withS(span.localServiceName() + "###" + span.name()));
+        }
+
+        if (span.remoteServiceName() != null && !span.remoteServiceName().isEmpty()) {
+          spanPut.addItemEntry(REMOTE_SERVICE_NAME_SPAN_NAME,
+              new AttributeValue().withS(span.remoteServiceName() + "###" + span.name()));
+        }
+      }
+
+      if (span.localServiceName() != null && !span.localServiceName().isEmpty()) {
+        spanPut.addItemEntry(LOCAL_SERVICE_NAME,
+            new AttributeValue().withS(span.localServiceName()));
+      }
+
+      if (span.remoteServiceName() != null && !span.remoteServiceName().isEmpty()) {
+        spanPut.addItemEntry(REMOTE_SERVICE_NAME,
+            new AttributeValue().withS(span.remoteServiceName()));
+      }
+
+      spanPut.addItemEntry(SPAN_BLOB,
+          new AttributeValue().withB(ByteBuffer.wrap(SpanBytesEncoder.PROTO3.encode(span))));
+
+      span.tags().forEach((key, value) -> {
+        if (value.isEmpty()) {
```

Comment:
No, it causes a 400 on the API call

Reply:
that sounds like a bug

Reply:
we have practice of "error" -> "" which I guess we have no explicit test in the codec tree for. will follow-up

Reply:
I mean that Dynamo doesn't allow values with an empty string, you have to explicitly use the NULL type to put no value

Reply:
As annoying as it is, but that's true, @adriancole. 

I can't imagine how many human effort was wasted on fixing bugs (sometimes in production) because of it.
Really weird decision from DynamoDB's engineers, but we have to deal with it...

Reply:
readback_minimalErrorSpan() passes.. I'm happy. thanks for the context!